File: python/day02/day02.py
import re

class Policy(object):
    def __init__(self, s):
        p = re.compile(r'(\d+)-(\d+) (.*)')
        groups = p.match(s).groups()
        if len(groups) != 3:
            raise Exception("Invalid policy string")
        self.position_one = int(groups[0])
        self.position_two = int(groups[1])
        self.character = groups[2]

class Password(object):
    def __init__(self, value, policy):
        self.value = value
        self.policy = policy
    
    def is_valid_one(self):
        count = list(self.value).count(self.policy.character)
        return count >= self.policy.position_one and count <= self.policy.position_two

    def is_valid_two(self):
        first = self.value[self.policy.position_one - 1] == self.policy.character
        second = self.value[self.policy.position_two - 1] == self.policy.character
        return first ^ second

def read_passwords(filename):
    fp = open(filename, 'r')
    lines = fp.readlines()
    fp.close()
    parts = list(map(lambda s: list(map(lambda c: c.strip(), s.split(":"))), lines))
    return [Password(p[1], Policy(p[0])) for p in parts]

print(len(list(filter(lambda p: p.is_valid_one(), read_passwords('input.txt')))))
print(len(list(filter(lambda p: p.is_valid_two(), read_passwords('input.txt')))))

File: ruby/day02/day02.rb
Policy = Struct.new(:position_one, :position_two, :character) do
  def self.from_string(s)
    matches = s.match(/(\d+)-(\d+) (.*)/)
    Policy.new(matches[1].to_i, matches[2].to_i, matches[3])
  end
end

Password = Struct.new(:password, :policy) do
  def self.from_string(s)
    parts = s.split(':').map(&:strip)
    Password.new(parts[1], Policy.from_string(parts[0]))
  end
end

def get_passwords
  lines = File.open('input.txt').readlines()
  lines.map { |l| Password.from_string(l) }
end

def is_valid(password, policy)
  count = password.split('').count(policy.character)
  count >= policy.position_one && count <= policy.position_two
end

def is_valid_two(password, policy)
  first = password[policy.position_one - 1] == policy.character
  second = password[policy.position_two - 1] == policy.character
  first ^ second
end

def valid_passwords(&validator)
  validated = get_passwords.map { |password| validator.call(password) }
  validated.count(true)
end

puts valid_passwords { |p| is_valid(p.password, p.policy) }
puts valid_passwords { |p| is_valid_two(p.password, p.policy) }